crates/chaincash_predicate/src/predicates/collateral.rs

```diff
@@ -0,0 +1,213 @@
+use crate::context::{ContextProvider, PredicateContext};
+use crate::predicates::Accept;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum CollateralAlgorithm {
+    Initial,
+}
+
+impl Default for CollateralAlgorithm {
+    fn default() -> Self {
+        Self::Initial
+    }
+}
+
+impl CollateralAlgorithm {
+    fn initial<P: ContextProvider>(&self, percent: u16, context: &PredicateContext<P>) -> bool {
+        let issuer_note_tally: u64 = context
+            .provider
+            .agent_issued_notes(&context.note.issuer)
+            .iter()
+            .map(|n| n.nanoerg)
+            .sum();
+        let issuer_reserves = context
+            .provider
+            .agent_reserves_nanoerg(&context.note.issuer);
+        let issuer_collateral = (issuer_reserves as f64 / issuer_note_tally as f64) * 100.0;
+
+        if issuer_collateral >= percent as f64 {
+            return true;
+        }
+
+        for signer in context.note.signers.iter().skip(1) {
+            let signer_notes = context.provider.agent_issued_notes(signer);
+            let highest_value_note = signer_notes.iter().max_by_key(|n| n.nanoerg);
+
+            if let Some(signer_note) = highest_value_note {
+                let signer_reserves = context.provider.agent_reserves_nanoerg(signer);
```

Comment:
I think we might need to get the reserves based on the reserves nft id in the signer chain in notes `r4`?